tools/onlineupdate.php

```diff
@@ -40,51 +47,56 @@ public function execute() {
-      if($match === FALSE)
-        $match = $this->CheckVersionResponse('alpha', $tiers, $platform, $InstalledVersion);
-      return $match;
-    }
-
-    private function CheckVersionResponse($tier, $tiers, $platform, $InstalledVersion) {
       if(!isset($tiers[$tier])) return FALSE;
       $tierdata = $tiers[$tier];
-      if($this->IsSameMajorVersion($tierdata->version, $InstalledVersion)) {
-        // TODO: Offer upgrades for MAJOR.x.x.x versions.
-        // We still don't support staying on alpha or beta tier once a version
-        // hits stable. We need to review the upgrade strategies for these.
-        // Once a version is older than latest stable, we also don't offer updates for it;
-        // this is probably also wrong.
-
-        $files = get_object_vars($tierdata->files);
-        foreach($files as $file => $filedata) {
-          // This is currently tied to Windows -- for other platforms we need to change this
-          if(preg_match($this->installerRegex, $file)) {
-            $filedata->url = \Keyman\Site\Common\KeymanHosts::Instance()->downloads_keyman_com . "/$platform/$tier/{$filedata->version}/{$file}";
-            return $filedata;
+
+      // Keyman Desktop version 10.0 should not automatically update to 13.0, because 11.0-13.0
+      // had a bug which broke keyboard registration details. Note, this is not yet fixed in 14.0
+      // either -- see keymanapp/keyman#3865.
+      if($platform == 'windows' && $this->IsSameMajorVersion($InstalledVersion, '10.0')) {
+        // For now, we don't even attempt to locate a newer 10.0 release.
+        return FALSE;
+      }
+
+      // Look for the file data that matches our requirements
+      $files = get_object_vars($tierdata->files);
+      foreach($files as $file => $filedata) {
+        if(preg_match($this->installerRegex, $file)) {
+          // We want to inject the final URL into the data returned from downloads.keyman.com
+          $filedata->url = KeymanHosts::Instance()->downloads_keyman_com . "/$platform/$tier/{$filedata->version}/{$file}";
+
+          if(!$this->IsSameMajorVersion($InstalledVersion, $tierdata->version)) {
+            // We're going to stagger upgrades by the minute of the hour for the check, to
+            // ensure we don't have everyone major-update at once and potentially cause us
+            // grief. This will mean that we need additional PRs to update this value; that
+            // gives us tracking automatically, so I'm good with that.
+
+            // For the initial rollout of this functionality, our stable release is 13.0, so
+            // we want to manually set the release date to around the time this PR lands.
+            $date = $this->IsSameMajorVersion('13.0', $tierdata->version) ? '2020-11-18' : $filedata->date;
+
+            if(!ReleaseSchedule::DoesRequestMeetSchedule($date)) {
+              return FALSE;
+            }
           }
+
+          return $filedata;
```

Comment:
This one doesn't seem to have an `isManual` check.  If that's a limitation due to the older version not specifying the option with its queries, I totally understand - this is mostly an observation, so do with it what you will.

Reply:
Since the GitHub's context for this comment is pretty poor...

```
  /**
   * Provides update checks for Keyman Desktop for versions 10.0-13.0.
   * For version 14.0 and later, see class WindowsUpdateCheck. As we
   * are providing update checks only for older, stable versions in this
   * class, there is no need to do checks on tier.
   */
```

Reply:
Yep, older versions don't pass that option.